feat(schema): Inference controls and improvements (#13028)
gatsbyjs/gatsby: docs/blog/2019-05-17-improvements-to-schema-customization/index.md

---

> title: Improvements to Schema Customization API - Available in Gatsby 2.4.0

Reviewer: These should all be changed from 2.4.0 to 2.5.0?

```suggestion
title: Improvements to Schema Customization API - Available in Gatsby 2.5.0
```

date: 2019-05-17
author: Mikhail Novikov
tags:
  - schema
  - graphql
---

Today we are releasing further improvements to the schema customization that [we've released in version 2.2.0](/blog/2019-03-18-releasing-new-schema-customization). You can use them with Gatsby 2.4.0.

It is now possible to indicate to Gatsby, that you want to add a resolver to an explicitly defined fields. Use extensions like `@link` and `@dateformat` to add default arguments or/and resolvers to fields. In addition, when `@dontInfer` is set, Gatsby will no longer run inference for marked type, allowing one to improve performance for large data sets.

## Summary

> After about a month of testing schema customization both here and in pre-release we determined a couple of issues. We set up to do this because we wanted to remove uncertainty in people's schemas, when the data changes. However, the original design allowed some uncertainties anyway. In addition, we have made inferrence a more heavy process, trading performance for consistency and didn't really provide a way to opt out of it completely. To summarize:

Reviewer: I took a go at expanding this to make it clearer that it's talking about issues with the 2.2.0 release of schema customization...

```suggestion
After about a month of testing schema customization both here and in pre-release we determined a couple of issues. The original aim of our schema customisation work was to remove uncertainty in people's schemas when their data changes. 

However, the original design allowed some uncertainties anyway. In addition, it made inference a more heavy process, trading performance for consistency without providing a way to opt out completely. To summarize, the schema customization work released in Gatsby 2.2.0 had the following issues:
```

- Resolvers and arguments of fields like Date and File was determined by inferred data
- There was no easy way to use arguments/resolvers to override the above
- Inferrence was run even when `@dontInfer` flag was on
- There was no way to control inference outside of SDL, eg in Type Builders

> Therefore we have some changes to the way we do inferrence. In addition, we are deprecating some of the features introduced is 2.2.0 and will remove them in Gatsby 3.

Reviewer: ```suggestion
Therefore we have some changes to the way we do inferrence. In addition, we are deprecating some of the features introduced in 2.2.0 and will remove them in Gatsby 3.
```

> 

Reviewer: Could add a heading here, something like "Updates in Gatsby 2.5.0"? Saying something about "these are the problems we identified, here are the updates we've created to fix them"

This would make it really clear where you've stopped talking about 2.2.0 problems and started talking about 2.5.0 updates.

Reviewer: Done in `f338bdd` 👍

## noDefaultResolvers and inferrence modes

> First of all, we are deprecating `noDefaultResolvers`. It was an argument of `infer` and `dontInfer`. We feel it was confusing and in some cases it didn't even actually add resolvers :). We will support `noDefaultResolvers` until version 3, after which `@infer` behaviour (see below) will become a default and `noDefaultResolvers` will be removed.

Reviewer: Some before and after code examples might be useful in these sections.

Reviewer: Done in `f338bdd` 👍

  define which field to compare to on a remote node and which field to use on
  the source node
- `@fileByRelativePath` - connect to a File node. Same arguments. The
  difference from link is that this normalizes the relative path to be
  relative from the path where source node is found.

```graphql
type MyType @infer {
  date: Date @dateformat(formatString: "DD MMM", locale: "fi")
  image: File @fileByRelativePath
  authorByEmail: Author @link(by: "email")
}
```

### Type Builders and extensions

You can now apply configuration to type builder types through extension property on them.

```js
schema.createObjectType({
  name: MyType,
  extensions: {
    infer: true,
  },
  fields: {
    date: {
      type: "Date",
      extensions: {
        dateformat: {
          formatString: "DD MMM",
          locale: "fi",
        },
      },
    },
  },
})
```

## Conclusions

With these improvements we hope we'll solve most of the issues that people are having with new schema customization. We are working on further improvements, like allowing users and plugins to define their own extensions (see [PR #13738](https://github.com/gatsbyjs/gatsby/pull/13738)).

> 

Reviewer: Could specifically ask for feedback and point out the umbrella issue link here?

Reviewer: Done in `f338bdd` 👍
